src/vm/memory.zig

```diff
@@ -104,71 +142,232 @@ pub const Region = struct {
         comptime state: MemoryState,
         vm_addr: u64,
         len: u64,
-    ) !state.Slice() {
-        if (vm_addr < self.vm_addr_start) return error.InvalidVirtualAddress;
+    ) ?state.Slice() {
+        if (vm_addr < self.vm_addr_start) return null;
 
-        const host_slice = try self.getSlice(state);
+        const host_slice = self.hostSlice(state) orelse return null;
         const begin_offset = vm_addr -| self.vm_addr_start;
-        if (try std.math.add(u64, begin_offset, len) <= host_slice.len) {
-            return host_slice[begin_offset..][0..len];
-        }
+        if (begin_offset +| len <= host_slice.len) return host_slice[begin_offset..][0..len];
+
+        return null;
+    }
+
+    fn lessThanFn(context: void, lhs: Region, rhs: Region) bool {
+        _ = context;
+        return lhs.vm_addr_start < rhs.vm_addr_start;
+    }
 
-        return error.VirtualAccessTooLong;
+    fn regionsOverlap(sorted_regions: []const Region) bool {
+        var iter = std.mem.window(Region, sorted_regions, 2, 1);
+
+        while (iter.next()) |region_pair|
+            if (region_pair[0].vm_addr_end > region_pair[1].vm_addr_start)
+                return true;
+
+        return false;
+    }
+
+    fn isValidAccess(self: Region, access_type: MemoryState, vm_addr: u64) bool {
+        if (access_type == .mutable and self.host_memory == .constant) return false;
+        if (vm_addr >= self.vm_addr_end) return false;
+        if (vm_addr < self.vm_addr_start) return false;
+
+        return true;
     }
 };
 
-const AlignedMemoryMap = struct {
+// [agave] https://github.com/anza-xyz/sbpf/blob/a8247dd30714ef286d26179771724b91b199151b/src/memory_region.rs#L551
+pub const AlignedMemoryMap = struct {
     regions: []const Region,
     version: sbpf.Version,
+    config: *const exe.Config,
 
-    fn init(regions: []const Region, version: sbpf.Version) !AlignedMemoryMap {
+    fn init(
+        regions: []const Region,
+        version: sbpf.Version,
+        config: *const exe.Config,
+    ) InitError!AlignedMemoryMap {
         for (regions, 1..) |reg, index| {
             if (reg.vm_addr_start >> VIRTUAL_ADDRESS_BITS != index) {
                 return error.InvalidMemoryRegion;
             }
         }
-
         return .{
             .regions = regions,
             .version = version,
+            .config = config,
         };
     }
 
-    fn region(self: *const AlignedMemoryMap, vm_addr: u64) !Region {
+    fn find_region(self: *const AlignedMemoryMap, vm_addr: u64) ?Region {
```

Comment:
here too

Reply:
changed to camelCase in e2a0e3c58e0f908854d59b9212eca28bc5f1bd0e